Panic when a package has been added to a catalog multiple times

When Grype is using a catalog that's had a package added more than once, Grype panics in [`FromCatalog`](https://github.com/anchore/grype/blob/0781fb028b56dbb4cc9ce1c0cb8988753e6f3b76/grype/pkg/package.go#L90-L96).

This is because Grype is adding items to the `result` slice by **index**, and it's trusting that the slice's length (currently set to the return value of `catalog.PackageCount()`) is large enough to accommodate any index value `i` that come from ranging over `catalog.Sorted()`.

This is not a safe assumption, since it's possible for `catalog.Sorted()` to return more items than the return value of `catalog.PackageCount()` indicates.

This issue was first discovered here: https://github.com/anchore/grype/issues/538#issuecomment-994890110 by @avelezxv

Example panic output:

```
panic: runtime error: index out of range [1216] with length 1216

goroutine 14 [running]:
github.com/anchore/grype/grype/pkg.FromCatalog(0xc004d16450, 0x12e4cf0, 0x8, 0xc004d16450)
        /Users/runner/work/grype/grype/grype/pkg/package.go:93 +0x272
github.com/anchore/grype/grype/pkg.syftProvider(0x7ffefe2e3b51, 0x10, 0x12e4cf0, 0x8, 0xc000c1b700, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/runner/work/grype/grype/grype/pkg/syft_provider.go:25 +0x15a
github.com/anchore/grype/grype/pkg.Provide(0x7ffefe2e3b51, 0x10, 0x12e4cf0, 0x8, 0xc000c1b700, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/runner/work/grype/grype/grype/pkg/provider.go:20 +0x115
github.com/anchore/grype/cmd.startWorker.func1.2(0xc0005623f0, 0x7ffefe2e3b51, 0x10, 0xc00056f458, 0xc00043bdf0, 0xc0002928b0, 0xc0001a21e0, 0xc0005623e7)
        /Users/runner/work/grype/grype/cmd/root.go:254 +0x105
created by github.com/anchore/grype/cmd.startWorker.func1
        /Users/runner/work/grype/grype/cmd/root.go:251 +0x35f
```

--- a/grype/pkg/package.go
+++ b/grype/pkg/package.go
@@ -88,9 +88,9 @@ func New(p pkg.Package) Package {
 }
 
 func FromCatalog(catalog *pkg.Catalog) []Package {
-	var result = make([]Package, catalog.PackageCount())
-	for i, p := range catalog.Sorted() {
-		result[i] = New(p)
+	result := make([]Package, 0, catalog.PackageCount())
+	for _, p := range catalog.Sorted() {
+		result = append(result, New(p))
 	}
 	return result
 }

--- a/grype/pkg/package_test.go
+++ b/grype/pkg/package_test.go
@@ -3,6 +3,8 @@ package pkg
 import (
 	"testing"
 
+	"github.com/anchore/syft/syft/source"
+
 	"github.com/anchore/syft/syft/file"
 	syftPkg "github.com/anchore/syft/syft/pkg"
 	"github.com/scylladb/go-set"
@@ -237,6 +239,27 @@ func TestNew_MetadataExtraction(t *testing.T) {
 			strset.Difference(observedMetadataTypes, expectedMetadataTypes),
 		)
 	}
+}
+
+func TestFromCatalog_DoesNotPanic(t *testing.T) {
+	catalog := syftPkg.NewCatalog()
+
+	examplePackage := syftPkg.Package{
+		Name:    "test",
+		Version: "1.2.3",
+		Locations: []source.Location{
+			source.NewLocation("/test-path"),
+		},
+		Type: syftPkg.NpmPkg,
+	}
+
+	catalog.Add(examplePackage)
+	// add it again!
+	catalog.Add(examplePackage)
+
+	assert.NotPanics(t, func() {
+		_ = FromCatalog(catalog)
+	})
 }
 
 func intRef(i int) *int {

